Mathlib/Tactic/Linter/DeprecateModule.lean

```diff
@@ -0,0 +1,138 @@
+  if yyyy.getNat < 2025 then
+    throwErrorAt yyyy "The year should be at least 2025!"
+  if mm.getNat == 0 || 12 < mm.getNat || mm.raw.getSubstring?.get!.toString.trim.length != 2 then
+    throwErrorAt mm "The month should be of the form 01, 02,..., 12!"
+  if dd.getNat == 0 || 31 < dd.getNat || dd.raw.getSubstring?.get!.toString.trim.length != 2 then
+    throwErrorAt dd "The day should be of the form 01, 02,..., 31!"
+  addModuleDeprecation
+  -- disable the linter, so that it does not complain in the file with the deprecation
+  elabCommand (← `(set_option linter.deprecateModule false))
+
+/-- A utility function to show the pairings `(deprecatedModule, #[preferredModules])`. -/
+elab "show_deprecated_modules" : command => do
+  let directImports := deprecateModuleExt.getState (← getEnv)
+  logInfo <| "\n".intercalate <|
+    directImports.fold (init := ["Deprecated modules\n"]) fun nms (i, deps) =>
+      nms ++ [s!"'{i}' deprecates to\n{deps}\n"]
+
+namespace DeprecateModule
+
+@[inherit_doc Mathlib.Linter.linter.deprecateModule]
+def deprecateModuleLinter : Linter where run := withSetOptionIn fun stx ↦ do
+  unless Linter.getLinterValue linter.deprecateModule (← getOptions) do
+    return
+  if (← get).messages.hasErrors then
+    return
+  let mainModule ← getMainModule
+  unless Linter.getLinterValue linter.style.header (← getOptions) do
+    return
+  if (← get).messages.hasErrors then
+    return
+  -- The imports of `Mathlib.lean` and `Mathlib.Tactic` are not required to be non-deprecated,
+  -- since these files are expected to import all modules satisfying certain path-restrictions.
+  if #[`Mathlib, `Mathlib.Tactic].contains mainModule then return
+  if stx.isOfKind ``Linter.deprecate_modules then return
+  let fm ← getFileMap
+  let md := (getMainModuleDoc (← getEnv)).toArray
+  -- The end of the first module doc-string, or the end of the file if there is none.
+  let firstDocModPos := match md[0]? with
+                          | none     => fm.positions.back!
+                          | some doc => fm.ofPosition doc.declarationRange.endPos
+  unless stx.getTailPos?.getD default ≤ firstDocModPos do
+    return
+  -- We try to parse the file up to `firstDocModPos`.
+  let upToStx ← parseUpToHere firstDocModPos <|> (do
+    -- If parsing failed, there is some command which is not a module docstring.
+    -- In that case, we parse until the end of the modules and add an extra `section` afterwards,
+    -- so we trigger a "no module doc-string" warning.
+    let fil ← getFileName
+    let (stx, _) ← Parser.parseHeader { input := fm.source, fileName := fil, fileMap := fm }
+    parseUpToHere (stx.getTailPos?.getD default) "\nsection")
+  let importIds := getImportIds upToStx
+  let modulesWithNames := importIds.map fun i => (i, i.getId)
+  let deprecations := deprecateModuleExt.getState (← getEnv)
+  for is@(i, _) in deprecations do
+    match modulesWithNames.find? (·.2 == i) with
```

Comment:
You could use `if let` here; probably saves two lines and one level of indentation.

Reply:
Actually: can you do `some (nmStx, _)` instead of the unpacking on line 129?

Reply:
I like this, thanks!